book/src/pil/types.md

```diff
@@ -9,13 +9,75 @@ The powdr-pil language has the following types:
 - tuple
 - array
 - function type
-- expression
-- constraint
+- `expr` (expression)
+- `constr` (constraint)
+- `!` ("bottom" or "unreachable" type)
+
+> The `col` type is special in that it is not a proper type, but can be used for declaring symbols.
+> See [Types of Symbols](#types-of-symbols) for details.
+
+Powdr-pil does perform a Hindley-Milner-kind type checking. After type checking, for all symbols
+without an explicit type, their type has to be uniquely determined. If you want a symbol to have
+a generic type, that generic type has to be explicitly specified.
+
+Declarations of generic symbols can constraint type variables to satisfy certain bounds.
+Currently, only built-in bounds or traits are supported (see the next section).
+
+Literal numbers do not have a specific type, they can be either `int`, `fe` or `expr` (the types that
+implement the `FromLiteral` trait), and their type can also stay generic until evaluation.
+
+### Example
+
+The following snippet defines a function that takes a value of a generic type and returns the value incremented by one.
+The type bounds on the generic type are `FromLiteral` and `Add`. The type checker will complain if we do not specify
+the type bounds. The bound `Add` is required because we use the `+` operator in the function and `FromLiteral` is needed
+because we use the literal `1` as a value of that type.
+
+```rust
+let<T: FromLiteral + Add> add_one: T -> T = |i| i + 1;
+```
+
+## Types of Symbols
+
+A symbol declared to have type `col` (or `col[k]`) is a bit special:
+
+If you assign it a value, that value is expected to have type `int -> fe` or `int -> int` (or an array thereof).
+This allows the simple declaration of a column `let byte = |i| i & 0xff;` without complicated conversions.
+The integer value is converted to a field element during evaluation, but it has to be non-negative and less than
+the field modulus.
+
+If you reference such a symbol, the type of the reference is `expr`, which means it is only symbolically referenced.
+A byte constraint is as easy as `{ X } in { byte }` now, since the expected types in plookup columns is `expr`.
+The downside is that you cannot evaluate columns as functions. If you want to do that, you have to assign
+a copy to an `int -> int` symbol: `let byte_f = |i| i & 0xff; let byte = byte_f;`.
```

Comment:
shouldn't `byte_f` and `byte` be explicitly typed?

Reply:
Yes, `byte` should be, `byte_f` does not need to be, but I can add types for clarity.